File: src/base/ServiceBase.ts
import { AxiosResponse, type AxiosRequestConfig } from "axios";
import { ApiClient } from "./ApiClient.js";
import { type SessionResponse } from "../models/responses/SessionResponse";
import { Response } from "../models/responses/Response";
import { CredentialsModel } from "../models/request/CredentialsModel.js";
import { SessionManager } from "./SessionManager.js";

export class ServiceBase {
  private api: ApiClient;
  protected url: string;
  protected defaultConfig: AxiosRequestConfig;

  constructor(endpointPath: string) {
    this.api = ApiClient.getInstance();
    this.url = this.baseUrl + endpointPath;
    this.defaultConfig = {};
  }

  get baseUrl(): string {
    return process.env["BASEURL"] ?? "";
  }

  async authenticate(): Promise<void> {
    const username = process.env["USER"];
    const password = process.env["PASSWORD"];

    if (!username || !password) {
      throw new Error("Missing username or password in environment variables.");
    }

    const cachedToken = SessionManager.getCachedToken(username, password);

    if (cachedToken) {
      this.defaultConfig = {
        headers: { Cookie: "token=" + cachedToken },
      };
      return;
    }

    const credentials: CredentialsModel = {
      username,
      password,
    };
    const response = await this.post<SessionResponse>(`${this.baseUrl}/auth`, credentials);

    SessionManager.storeToken(username, password, response.data.token);

    this.defaultConfig = {
      headers: { Cookie: "token=" + response.data.token },
    };
  }

  protected async get<T>(
    url: string,
    config: AxiosRequestConfig = this.defaultConfig,
  ): Promise<Response<T>> {
    const startTime = Date.now();
    const response = await this.api.client.get<T>(url, config);
    const endTime = Date.now();

    const customResponse: Response<T> = this.buildResponse<T>(endTime, startTime, response);
    return customResponse;
  }

  protected async post<T>(
    url: string,
    data: unknown,
    config: AxiosRequestConfig = this.defaultConfig,
  ): Promise<Response<T>> {
    const startTime = Date.now();
    const response = await this.api.client.post<T>(url, data, config);
    const endTime = Date.now();

    const customResponse: Response<T> = this.buildResponse<T>(endTime, startTime, response);
    return customResponse;
  }

  protected async put<T>(
    url: string,
    data: unknown,
    config: AxiosRequestConfig = this.defaultConfig,
  ): Promise<Response<T>> {
    const startTime = Date.now();
    const response = await this.api.client.put<T>(url, data, config);
    const endTime = Date.now();

    const customResponse: Response<T> = this.buildResponse<T>(endTime, startTime, response);
    return customResponse;
  }

  protected async patch<T>(
    url: string,
    data: unknown,
    config: AxiosRequestConfig = this.defaultConfig,
  ): Promise<Response<T>> {
    const startTime = Date.now();
    const response = await this.api.client.patch<T>(url, data, config);
    const endTime = Date.now();

    const customResponse: Response<T> = this.buildResponse<T>(endTime, startTime, response);
    return customResponse;
  }

  protected async delete<T>(
    url: string,
    config: AxiosRequestConfig = this.defaultConfig,
  ): Promise<Response<T>> {
    const startTime = Date.now();
    const response = await this.api.client.delete<T>(url, config);
    const endTime = Date.now();

    const customResponse: Response<T> = this.buildResponse<T>(endTime, startTime, response);
    return customResponse;
  }

  protected async head<T>(
    url: string,
    config: AxiosRequestConfig = this.defaultConfig,
  ): Promise<Response<T>> {
    const startTime = Date.now();
    const response = await this.api.client.head<T>(url, config);
    const endTime = Date.now();

    const customResponse: Response<T> = this.buildResponse<T>(endTime, startTime, response);
    return customResponse;
  }

  protected async options<T>(
    url: string,
    config: AxiosRequestConfig = this.defaultConfig,
  ): Promise<Response<T>> {
    const startTime = Date.now();
    const response = await this.api.client.options<T>(url, config);
    const endTime = Date.now();

    const customResponse: Response<T> = this.buildResponse<T>(endTime, startTime, response);
    return customResponse;
  }

  private buildResponse<T>(endTime: number, startTime: number, response: AxiosResponse<T>) {
    const responseTime = endTime - startTime;

    const customResponse: Response<T> = {
      data: response.data,
      status: response.status,
      headers: response.headers,
      responseTime: responseTime,
    };
    return customResponse;
  }
}

File: src/base/ServiceBase.py
import os
from src.base.ApiClient import ApiClient
from src.base.SessionManager import SessionManager
from src.models.request.CredentialsModel import CredentialsModel
from src.models.responses.Response import Response

class ServiceBase:
    def __init__(self, endpoint_path: str):
        self.api = ApiClient()
        self.url = self.base_url + endpoint_path
        self.default_config = {}

    @property
    def base_url(self) -> str:
        return os.getenv("BASEURL", "")

    def authenticate(self) -> None:
        username = os.getenv("USER")
        password = os.getenv("PASSWORD")

        if not username or not password:
            raise ValueError("Missing username or password in environment variables.")

        cached_token = SessionManager.get_cached_token(username, password)

        if cached_token:
            self.default_config = {
                "headers": {"Cookie": "token=" + cached_token},
            }
            return

        credentials = CredentialsModel(username=username, password=password)
        response = self.post(f"{self.base_url}/auth", credentials)

        SessionManager.store_token(username, password, response.data.token)

        self.default_config = {
            "headers": {"Cookie": "token=" + response.data.token},
        }

    def get(self, url: str, config: dict = None) -> Response:
        if config is None:
            config = self.default_config
        start_time = self._get_current_time()
        response = self.api.client.get(url, **config)
        end_time = self._get_current_time()
        return self._build_response(end_time, start_time, response)

    def post(self, url: str, data: dict, config: dict = None) -> Response:
        if config is None:
            config = self.default_config
        start_time = self._get_current_time()
        response = self.api.client.post(url, json=data, **config)
        end_time = self._get_current_time()
        return self._build_response(end_time, start_time, response)

    def put(self, url: str, data: dict, config: dict = None) -> Response:
        if config is None:
            config = self.default_config
        start_time = self._get_current_time()
        response = self.api.client.put(url, json=data, **config)
        end_time = self._get_current_time()
        return self._build_response(end_time, start_time, response)

    def patch(self, url: str, data: dict, config: dict = None) -> Response:
        if config is None:
            config = self.default_config
        start_time = self._get_current_time()
        response = self.api.client.patch(url, json=data, **config)
        end_time = self._get_current_time()
        return self._build_response(end_time, start_time, response)

    def delete(self, url: str, config: dict = None) -> Response:
        if config is None:
            config = self.default_config
        start_time = self._get_current_time()
        response = self.api.client.delete(url, **config)
        end_time = self._get_current_time()
        return self._build_response(end_time, start_time, response)

    def head(self, url: str, config: dict = None) -> Response:
        if config is None:
            config = self.default_config
        start_time = self._get_current_time()
        response = self.api.client.head(url, **config)
        end_time = self._get_current_time()
        return self._build_response(end_time, start_time, response)

    def options(self, url: str, config: dict = None) -> Response:
        if config is None:
            config = self.default_config
        start_time = self._get_current_time()
        response = self.api.client.options(url, **config)
        end_time = self._get_current_time()
        return self._build_response(end_time, start_time, response)

    def _build_response(self, end_time: int, start_time: int, response) -> Response:
        response_time = end_time - start_time
        return Response(
            data=response.json(),
            status=response.status_code,
            headers=response.headers,
            response_time=response_time,
        )

    def _get_current_time(self) -> int:
        from time import time
        return int(time() * 1000)
